# CONTENT LIST
01. Importing Libraries
02. Importing Data
03. Data Consistency Checks
04. Beginning of task 5

# 1.Importing Libraries

In [46]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [47]:
# Creating common folder path
path = r'C:\Users\Mukund\Desktop\Career Foundry\Instacart_Grocery_Basket_Analysis_May_2021'

In [48]:
path

'C:\\Users\\Mukund\\Desktop\\Career Foundry\\Instacart_Grocery_Basket_Analysis_May_2021'

In [49]:
# Importing orders data
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared_data', 'orders_wrangled.csv'), index_col = False)

In [50]:
# Importing products data
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original_data', 'products.csv'), index_col = False)

In [10]:
# Creating test dataframe
df_test = pd.DataFrame()

In [11]:
# Creating mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Checking for mixed-type data in orders
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# 3.Data Consistency Checks

In [14]:
# Checking for number of missing values in products
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prods.shape

(49693, 5)

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [74]:
# Exporting df_prods dataframe under Prepared Data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared_data', 'Products_Checked.csv'), index = False)

# 4. Beginning of task 5

Step 1) If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now. - Done above

Step 2) Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [32]:
# Checking for the data using describe function
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The max of price column is very high 99999 which can be a mistake or human error while entering data. Also, count of the product_id column does not match with the max of product_id

Step 3) Check for mixed-type data in your df_ords dataframe.

In [51]:
# Checking for mixed-type data from orders
for col in df_ords.columns.tolist():
  mixed_type = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[mixed_type]) > 0:
    print (col)

No mixed-type data found

Step 5) Run a check for missing values in your df_ords dataframe

In [64]:
# Checking for missing values in orders data
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

Only one column contains the missing data which is 'days_since_prior_order' This can be possible if customer is ordering for the first time.

Step 6) Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [57]:
# Getting the number of records for orders data
df_ords.shape

(3421083, 6)

First of all, the missing data is not the wrong one because there can be possibility of ordering for first time and there can be no records or days available in that column. So, first I will confirm with the client whether to remove this data or to impute mean/median or any other value. As of now I am going to impute 0 values for the empty records.

In [58]:
# Getting the data where 'days_since_prior_order' column not having missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [59]:
df_nan_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [60]:
# Updating empty records with 0 value
df_ords['days_since_prior_order'].fillna(0, inplace = True)

In [65]:
df_ords.shape

(3421083, 6)

In [67]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [ ]:
No missing data found after updating

Step 7) Run a check for duplicate values in your df_ords data.

In [68]:
# Checking for duplicates in orders data
df_ords_dups = df_ords[df_ords.duplicated()]

In [69]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates found in orders data

In [72]:
#Exporting df_ords dataframe under Prepared Data folder
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared_data', 'Orders_Checked.csv'),index=False)

In [73]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]